Content.Client/DragDrop/DragDropSystem.cs

```diff
@@ -18,479 +16,525 @@
+        {
+            return false;
         }
 
-        private bool OnBeginDrag()
+        if (_dragShadow == null)
+            return false;
+
+        _targetRecheckTime += frameTime;
+
+        if (_targetRecheckTime > TargetRecheckInterval)
         {
-            if (_dragDropHelper.Dragged == default || Deleted(_dragDropHelper.Dragged))
-            {
-                // something happened to the clicked entity or we moved the mouse off the target so
-                // we shouldn't replay the original click
-                return false;
-            }
+            HighlightTargets();
+            _targetRecheckTime -= TargetRecheckInterval;
+        }
 
-            if (EntityManager.TryGetComponent<SpriteComponent?>(_dragDropHelper.Dragged, out var draggedSprite))
+        return true;
+    }
+
+    private bool OnUseMouseUp(in PointerInputCmdHandler.PointerInputCmdArgs args)
+    {
+        if (_state == DragState.MouseDown)
+        {
+            // haven't started the drag yet, quick mouseup, definitely treat it as a normal click by
+            // replaying the original cmd
+            if (_savedMouseDown.HasValue && _mouseDownTime < MaxMouseDownTimeForReplayingClick)
             {
-                // pop up drag shadow under mouse
-                var mousePos = _eyeManager.ScreenToMap(_dragDropHelper.MouseScreenPosition);
-                _dragShadow = EntityManager.SpawnEntity("dragshadow", mousePos);
-                var dragSprite = EntityManager.GetComponent<SpriteComponent>(_dragShadow);
-                dragSprite.CopyFrom(draggedSprite);
-                dragSprite.RenderOrder = EntityManager.CurrentTick.Value;
-                dragSprite.Color = dragSprite.Color.WithAlpha(0.7f);
-                // keep it on top of everything
-                dragSprite.DrawDepth = (int) DrawDepth.Overlays;
-                if (!dragSprite.NoRotation)
+                var savedValue = _savedMouseDown.Value;
+                _isReplaying = true;
+                // adjust the timing info based on the current tick so it appears as if it happened now
+                var replayMsg = savedValue.OriginalMessage;
+                var adjustedInputMsg = new FullInputCmdMessage(args.OriginalMessage.Tick, args.OriginalMessage.SubTick,
+                    replayMsg.InputFunctionId, replayMsg.State, replayMsg.Coordinates, replayMsg.ScreenCoordinates, replayMsg.Uid);
+
+                if (savedValue.Session != null)
                 {
-                    EntityManager.GetComponent<TransformComponent>(_dragShadow).WorldRotation = EntityManager.GetComponent<TransformComponent>(_dragDropHelper.Dragged).WorldRotation;
+                    _inputSystem.HandleInputCommand(savedValue.Session, EngineKeyFunctions.Use, adjustedInputMsg, true);
                 }
 
-                HighlightTargets();
-                _outline.SetEnabled(false);
-
-                // drag initiated
-                return true;
+                _isReplaying = false;
```

Comment:
I guess this should do a try-finally to prevent dragging from being bricked if the input handling throws some exceptions?